Use Integers in GetNextPixelBoundaryTimeNs

karupayun opened the pull request:
* Use Integers in GetNextPixelBoundaryTimeNs

We use the function GetNextPixelBoundaryTimeNs to get the next time to
draw, in order to not draw a timer which all its pixels have already
another element drawn.

This function have 2 issues which leads to return sometimes a previous
time (and therefore incorrect one), which is currently making us drawing
timers than otherwise we will skip.

The first issue is that we have an integer value for ns_per_pixel, which
leads us to a problem because if we multiply the number of pixel with
ns_per_pixel we don't have the total number of pixels. We will fix this
in a different PR. The second issue is that although we use a double
there, we will have precision issues. So, for that we will make all our
calculations using integers, to avoid that.

Addressing: http://b/199736246

Test: Navigate in a capture, confirm that always the function answer a
reasonable time. Load a capture, pixelwise we seem to draw more elements, so we seems to be more 
precise.

Before: https://screenshot.googleplex.com/76KYwkkjk7c3eup
After: https://screenshot.googleplex.com/4WEjtteXDethqMz

pierricgimmig commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -446,11 +446,17 @@ void ThreadTrack::OnCaptureComplete() {
+  uint64_t total_ns = draw_data.max_tick - draw_data.min_tick;
+  uint64_t num_pixels_on_track = draw_data.viewport->WorldToScreenWidth(draw_data.track_width);
+
+  // Making a ceiling with integers to not lose precision. Giving a maximum number of 4000 pixels
```
"Given a track width of 4000 pixels, we can capture for more than a month without overflowing."

Orbit could span multiple screens, so it's better to mention the track width. Also, "giving" -> "given".

pierricgimmig replied:
Also, I had to validate that claim, my calculations bring me to ~53 days :-), I like the conservatism.

karupayun replied:
I also arrived to the same number. I will use the days instead to be more precise.

pierricgimmig reviewed (approved):
Coool, thanks for this Pablo! LGTM

vickyliu-go4it commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -446,11 +446,17 @@ void ThreadTrack::OnCaptureComplete() {
+
+  // Making a ceiling with integers to not lose precision. Giving a maximum number of 4000 pixels
+  // width in a screen, we can capture for more than a month without having overflow.
+  uint64_t next_pixel = ((ns_from_start * num_pixels_on_track) + total_ns - 1) / total_ns;
+
+  uint64_t next_pixel_ns_from_start = total_ns * next_pixel / num_pixels_on_track;
```
On the above line, we divide the value by `total_ns` but in the next line, we multiply the value by `total_ns`.
It's a good idea for improving the readability. But as here our may concern is the precision, I prefer to only have `next pixel` in the comments, for example:

```cpp
// We compute next_pixel_ns_from_start by using:
//     uint64_t next_pixel = ((ns_from_start * num_pixels_on_track) + total_ns - 1) / total_ns;
//     uint64_t next_pixel_ns_from_start = total_ns * next_pixel / num_pixels_on_track;
 uint64_t next_pixel_ns_from_start = ((ns_from_start * num_pixels_on_track) + total_ns - 1) / num_pixels_on_track;
```

karupayun replied:
As discussed offline, every step was part of being precise. 

The first division is the way to make a ceiling in integers:
ceiling(A / B) = (A + B - 1) / B;

I've added a comment below to explain why we are multiplying again. I think it should be enough.

vickyliu-go4it replied:
If that is the case, we can use ceiling(A/B) = 1 + (A-1) / B to avoid the overflowing of (A+B), as long as A != 0

karupayun replied:
Thanks for the suggestion. A should be != 0 and with your proposal we will gain a little more time (between 20 and 30 minutes), but I will leave as it is because I don't think that we could reach that time and I'm used to read this formula for ceiling, so it's easier to read for me at least.

vickyliu-go4it commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -446,11 +446,17 @@ void ThreadTrack::OnCaptureComplete() {
-  int pixel_x = static_cast<int>(
-      ceil(normalized_x * draw_data.viewport->WorldToScreenWidth(draw_data.track_width)));
-  return draw_data.min_tick + pixel_x * draw_data.ns_per_pixel;
+    uint64_t time_ns, const internal::DrawData& draw_data) {
```
Could I ask for a more descriptive name for `time_ns`? XD

karupayun replied:
I've renamed all the variables, let me know what you think.

vickyliu-go4it replied:
LGTM!

vickyliu-go4it reviewed (approved):
Nice job! Only some minor suggestions XD

karupayun commented:
Hey @pierricgimmig, @vickyliu-go4it. I've modified the comments, renaming variables but I also added a check for when we are zooming-in so much and end_tick == start_tick. Orbit works very strange at this case and we should fix it, but I don't want to crash here because a division by 0.

Re-asking for a review.

pierricgimmig commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -446,11 +446,25 @@ void ThreadTrack::OnCaptureComplete() {
+  uint64_t total_ns = draw_data.max_tick - draw_data.min_tick;
+  uint64_t num_pixels_on_track = draw_data.viewport->WorldToScreenWidth(draw_data.track_width);
+
+  // If max_tick and min_tick are the same, there is no way to calculate the next pixel, as every
+  // pixel has the same timestamp.
+  if (total_ns == 0) {
+    return current_end_time + 1;
+  }
```
Shouldn't this return draw_data.max_tick? If all pixels have the same timestamp, any timer for that timestamp will be drawn on the first iteration, we shouldn't draw anything else, no?

karupayun replied:
Exactly. Currently the UI is completely broken (for example see http://b/168283058) but it doesn't crash at all. I'm not sure if we should fix or limit the zooming-level only to not have to address this border cases.

In this case, we probably will be inside the only timer that we should draw:
current_start_time <= min_tick = max_tick <= current_end_time, and we should draw only that timer.

The problem of returning max_tick is that we are returning a pixel to draw inside the drawable timer, so basically we are telling that we can draw this timer again. With my optimization this will result in an infinite loop, because I have a implicit invariant that this function should return a time greater than the current_end_time provided.

vickyliu-go4it commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -457,12 +457,12 @@ void ThreadTrack::OnCaptureComplete() {
-  // can capture for 53 days without overflowing.
-  int next_pixel = ((ns_from_min * num_pixels_on_track) + total_ns - 1) / total_ns;
+  // Given a track width of 4000 pixels, we can capture for 53 days without overflowing.
+  int current_pixel = (ns_from_min * num_pixels_on_track) / total_ns;
```
You had `ns_from_min * num_pixels_on_track - 1` before, but now `-1` is gone.
Is this intended?

karupayun replied:
Yes, it's intended. We changed the logic there because it was wrong when we were exactly in the boundary. Now should be easy to understand.

karupayun commented:
I will merge this finally, with the current modifications seems to be that pixelwise we are making a better job, but it's only an intuition based on a few captures where for sure there are less blank spaces.
